scylla-cql/src/types/deserialize/value.rs

```diff
@@ -0,0 +1,1515 @@
+        let v = ensure_not_null_slice(v)?;
+        let mut mem = v.as_slice();
+        let count = types::read_int_length(&mut mem)?;
+        let (k_typ, v_typ) = match typ {
+            ColumnType::Map(k_t, v_t) => (k_t, v_t),
+            _ => {
+                return Err(ParseError::BadIncomingData(format!(
+                    "Expected map, got {:?}",
+                    typ,
+                )))
+            }
+        };
+        Ok(Self {
+            k_typ,
+            v_typ,
+            raw_iter: FixedLengthBytesSequenceIterator::new(
+                2 * count,
+                FrameSlice::new_subslice(mem, v.as_bytes_ref()),
+            ),
+            phantom_data_k: std::marker::PhantomData,
+            phantom_data_v: std::marker::PhantomData,
+        })
+    }
+}
+
+impl<'frame, K, V> Iterator for MapIterator<'frame, K, V>
+where
+    K: DeserializeCql<'frame>,
+    V: DeserializeCql<'frame>,
+{
+    type Item = Result<(K, V), ParseError>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let raw_k = match self.raw_iter.next() {
+            Some(Ok(raw_k)) => raw_k,
+            Some(Err(err)) => return Some(Err(err)),
+            None => return None,
+        };
+        let raw_v = match self.raw_iter.next() {
+            Some(Ok(raw_v)) => raw_v,
+            Some(Err(err)) => return Some(Err(err)),
+            None => return None,
+        };
+        let do_next = || -> Result<(K, V), ParseError> {
+            let k = K::deserialize(self.k_typ, raw_k)?;
+            let v = V::deserialize(self.v_typ, raw_v)?;
+            Ok((k, v))
+        };
+        do_next().map(Some).transpose()
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        self.raw_iter.size_hint()
+    }
+}
+
+impl<'frame, K, V> DeserializeCql<'frame> for BTreeMap<K, V>
+where
+    K: DeserializeCql<'frame> + Ord,
+    V: DeserializeCql<'frame>,
+{
+    fn type_check(typ: &ColumnType) -> Result<(), ParseError> {
+        MapIterator::<'frame, K, V>::type_check(typ)
+    }
+
+    fn deserialize(
+        typ: &'frame ColumnType,
+        v: Option<FrameSlice<'frame>>,
+    ) -> Result<Self, ParseError> {
+        MapIterator::<'frame, K, V>::deserialize(typ, v)?.collect()
+    }
+}
+
+impl<'frame, K, V, S> DeserializeCql<'frame> for HashMap<K, V, S>
+where
+    K: DeserializeCql<'frame> + Eq + Hash,
+    V: DeserializeCql<'frame>,
+    S: BuildHasher + Default + 'frame,
+{
+    fn type_check(typ: &ColumnType) -> Result<(), ParseError> {
+        MapIterator::<'frame, K, V>::type_check(typ)
+    }
+
+    fn deserialize(
+        typ: &'frame ColumnType,
+        v: Option<FrameSlice<'frame>>,
+    ) -> Result<Self, ParseError> {
+        MapIterator::<'frame, K, V>::deserialize(typ, v)?.collect()
+    }
+}
+
+// tuples
+
+// Implements tuple deserialization.
+// The generated impl expects that the serialized data will contain at least
+// the given amount of values.
+// TODO: Add more context to errors
```

Comment:
What do you mean here? Please specify in the comment.

Reply:
Which part of the comment is unclear?

Reply:
Clarified: `TODO: Include information about the id of the column that failed to parse`